### ISSUE TITLE
Recovery of volume with trailing data and file mark on data partition breaks index traversal

### ISSUE THREAD

**Issue by linnemannr** [LinearTapeFileSystem/ltfs#272]
**Description**

When data is written to an LTFS volume, but not committed with an index, recovery of that volume with ltfsck does not take into account whether the trailing data blocks were terminated with a file mark. As a result, when ltfsck restores consistency to a volume by re-writing the index referenced in the VCI to the end of the partition, two consecutive file marks follow the trailing data blocks. This breaks both forward and backward index traversal. Traversing forward, libltfs sees an empty index document after the original consistent index, and traversing backward it sees the trailing data where it expects to have an XML document to parse.

**To Reproduce**

1. Create a tape with the file backend driver
2. In the file backend tape directory, emulate the condition of trailing data with a file marker:
    ```
    $ rm 1_7_E
    $ echo "trailing data" > 1_7_R
    $ touch 1_8_F 1_9_E```
3. Recover the volume with ltfsck:
    ```
    $ ltfsck -d ./path/to/my/tape
    ```
4. Attempt to list the rollback points with ltfsck:
    ```
    $ ltfsck -d ./path/to/my/tape -l
    ```
5. ltfsck will not be able to traverse backward beyond the newly written generation 2:
    ```Generation: Date       Time               Zone     SelfPtr->BackPtr (Part, Pos)
               (UTC Date   UTC Time           UTC)
                Commit Message
    -------------------------------------------------------------------------------
             2: 2021-04-19 22:22:35.499777196 UTC      (0, 5)->(1, 10)
                No commit message
             2: 2021-04-19 22:22:35.499777196 UTC      (1, 10)->(1, 5)
                No commit message
    Entity: line 1: parser error : Document is empty
    trailing data
    ^
    LTFS17037E XML parser: failed to read from XML stream.
    LTFS17016E Cannot parse index direct from medium.
    LTFS11194W Cannot read index: failed to read and parse XML data (-1014).
    LTFS17075E Traverse(B): cannot read index from 8 (b).
    LTFS16076E Cannot list rollback points: failed to traverse the data partition (-1014).
    ```

    Traversing forward, ltfsck will not be able to traverse beyond generation 1:
    ```Generation: Date       Time               Zone     SelfPtr->BackPtr (Part, Pos)
               (UTC Date   UTC Time           UTC)
                Commit Message
    -------------------------------------------------------------------------------
             2: 2021-04-19 22:22:35.499777196 UTC      (0, 5)->(1, 10)
                No commit message
             1: 2021-04-19 22:21:02.023604489 UTC      (1, 5) <<Initial Index>>
                No commit message
    Entity: line 1: parser error : Extra content at the end of the document
    
    ^
    LTFS17037E XML parser: failed to read from XML stream.
    LTFS17016E Cannot parse index direct from medium.
    LTFS11194W Cannot read index: failed to read and parse XML data (-1014).
    LTFS17075E Traverse(F): cannot read index from 9 (b).
    LTFS16076E Cannot list rollback points: failed to traverse the data partition (-1014).
    ```

**Expected behavior**

ltfs_check_medium() should note whether the data partition ends with trailing blocks terminated with a filemark. Since ltfs_write_index() writes both filemarks that fence the index, the append position should be set to the block  location just after the trailing data.

**Additional context**

This problem is especially likely to occur on FreeBSD with the cam driver, where closing of the sa() device after a write automatically writes a file mark. If the ltfs application (or another application linked to and using libltfs)  terminates after writing data blocks without committing an index, the partition exhibits the conditions for the recovery to lay down the additional file mark after the trailing data, making previous generations of the volume inaccessible via ltfsck.

If the volume policy allows data in the index partition, this situation could arise there as well.

**Comment by piste-jp-ibm** [comment]
The list looks funny...

The list saids an index is started from (1, 10) and the backpointer points (0, 5). On the other hand you deleted the end marker on (1, 7). Why can `ltfsck` parse the index on (1, 10) ?

**Comment by linnemannr** [comment]
I moved the EOD to block 1,9. Block 1,7 was made to be a block of data
trailing the index at 1,5. A file mark was added to the end of the data at
1,8. ltfsck attempted to recover the inconsistent volume by loading the
index from 1,5, marking the volume dirty, and then committing a new
generation to the index and data partitions. On the data partition, the
leading file mark landed at the original EOD 1,9, the index at 1,10, and
the trailing file mark at 1,11. The current EOD is at 1,12. The result is a
consistent volume with EOD, however the back-to-back file marks at blocks
1,8 and 1,9 violate the specification.
